bottomless2/src/scheduler.rs

```diff
@@ -13,44 +30,342 @@ use crate::{
 /// A new segment can not be enqueued until the previous segment for the same namespace has been
 /// processed, because only the most recent segment is checked for durability. This property
 /// ensures that all segments are present up to the max durable index.
-pub(crate) struct Scheduler<S> {
+pub(crate) struct Scheduler<S: Storage> {
     capacity: usize,
     /// notify new durability index for namespace
-    durable_notifier: mpsc::Sender<(libsql_wal::name::NamespaceName, u64)>,
-    storage: S,
+    durable_notifier: mpsc::Sender<(NamespaceName, u64)>,
+    requests: HashMap<NamespaceName, NamespaceRequests<S::Config>>,
+    queue: priority_queue::PriorityQueue<NamespaceName, Reverse<u64>>,
+    storage: Arc<S>,
+    next_request_id: u64,
 }
 
 impl<S: Storage> Scheduler<S> {
+    pub fn new(
+        capacity: usize,
+        durable_notifier: mpsc::Sender<(NamespaceName, u64)>,
+        storage: S,
+    ) -> Self {
+        Self {
+            capacity,
+            durable_notifier,
+            requests: Default::default(),
+            queue: Default::default(),
+            storage: Arc::new(storage),
+            next_request_id: Default::default(),
+        }
+    }
+
     /// Register a new request with the scheduler
+    #[tracing::instrument(skip_all)]
     pub fn register(&mut self, request: StoreSegmentRequest<S::Config>) {
         // invariant: new segment comes immediately after the latest segment for that namespace. This means:
         // - immediately after the last registered segment, if there is any
         // - immediately after the last durable index
-        todo!()
+        let id = self.next_request_id;
+        self.next_request_id += 1;
+        let name = request.namespace.clone();
+        let slot = IndexedRequest { request, id };
+        let requests = self.requests.entry(name.clone()).or_default();
+        requests.requests.push_back(slot);
+
+        tracing::debug!(job_id = id, "job registered");
+
+        // if there is a priority for this namespace already, it must be higher than ours, because
```

Comment:
Can you expand on this comment, I think this logic is a bit confusing and would be good to have a clarifying comment.

Reply:
it's quite intuitive really: if there's already some work in the queue for that namespace, then it must have been registered before, and if it was registered before, then it must be processed before, so we don't update the priority for that namespace by inserting a more recent job, because that would decrease the priority of the namespace